File: src/tools/fund_wallet.py
from typing import Optional, Type
from json import dumps
from langchain.pydantic_v1 import BaseModel, Field
from langchain_core.callbacks import CallbackManagerForToolRun
from langchain_core.tools import BaseTool
from coinbase_sdk.api import CoinbaseAPIWrapper, DEFAULT_WALLET_ID
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class FundWalletInput(BaseModel):
    wallet_id: str = Field(description="The wallet ID to fund with testnet ETH")

class FundWalletTool(BaseTool):
    name = "FundWallet"
    description = "Fund a wallet with testnet ETH"
    args_schema: Type[BaseModel] = FundWalletInput
    return_direct: bool = True
    api: CoinbaseAPIWrapper = None

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.api = CoinbaseAPIWrapper()  # Initialize our API wrapper

    def _run(
        self,
        wallet_id: str,
        run_manager: Optional[CallbackManagerForToolRun] = None,
        **kwargs  # Absorbs any unexpected keyword arguments
    ) -> str:
        try:
            logger.info(f"Funding wallet with ID: {wallet_id}")
            result = self.api.fund_wallet(wallet_id)
            logger.info(f"Wallet funded successfully: {result}")
            return dumps(result)
        except Exception as e:
            logger.error(f"Funding failed for wallet ID {wallet_id}: {str(e)}")
            return f"Funding failed: {str(e)}"

# Usage example:
if __name__ == "__main__":
    tool = FundWalletTool()
    result = tool._run(wallet_id=DEFAULT_WALLET_ID)
    print(result)

File: src/tools/fundWallet.js
const { BaseModel, Field } = require('langchain/pydantic_v1');
const { BaseTool } = require('langchain_core/tools');
const { Coinbase } = require('@coinbase/coinbase-sdk');

class FundWalletInput extends BaseModel {
  constructor(data) {
    super(data);
    this.walletId = Field(data.walletId, { description: "The ID of the wallet to fund with testnet ETH" });
  }
}

class FundWalletTool extends BaseTool {
  constructor() {
    super();
    this.name = "FundWallet";
    this.description = "Fund a wallet with testnet ETH";
    this.argsSchema = FundWalletInput;
    this.returnDirect = true;
    this.coinbase = null;
  }

  async initialize() {
    try {
      this.coinbase = await Coinbase.configureFromJson({ 
        filePath: '~/Downloads/cdp_api_key.json' // Update with your API key file path
      });
      console.log('Coinbase SDK initialized successfully');
    } catch (error) {
      console.error('Failed to initialize Coinbase SDK:', error);
      throw error;
    }
  }

  async _run(walletId, runManager = null) {
    if (!this.coinbase) {
      await this.initialize();
    }

    try {
      const user = await this.coinbase.getDefaultUser();
      const wallet = await user.getWallet(walletId);
      const faucetTransaction = await wallet.faucet();
      return `Wallet funded successfully: ${JSON.stringify(faucetTransaction)}`;
    } catch (error) {
      console.error('Error funding wallet:', error);
      return `Funding failed: ${error.message}`;
    }
  }
}

// Usage example:
async function main() {
  const tool = new FundWalletTool();
  await tool.initialize();
  const result = await tool._run("wallet_id_here");
  console.log(result);
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = { FundWalletTool };